File: benchmark/2026-09-17-7677581(latest)/micro/src/java/mb06.java
// MB-06 - Matrix multiplication: classical i-j-k product over flat row-major arrays.
import java.util.Locale;

class Main {

    // Lehmer / MINSTD generator, frozen for every language in the suite.
    static final class Lcg {
        private long state;

        Lcg(long seed) {
            this.state = seed;
        }

        long nextInt() {
            state = (48271L * state) % 2147483647L;
            return state;
        }

        double nextUnit() {
            return nextInt() / 2147483647.0;
        }
    }

    private static final int N = 512;
    private static final int R = 3;
    private static final long SEED = 20266917L;

    /** The reported fields of one whole workload execution. */
    record Result(double sumC, double cFirst, double cLast) { }

    /** The whole workload body (methodology 06 section 5.2): re-seeds and regenerates A and B. */
    static Result workload() {
        Lcg gen = new Lcg(SEED);
        double[] a = new double[N * N];
        double[] b = new double[N * N];
        double[] c = new double[N * N];
        for (int i = 0; i < N * N; i++) {
            a[i] = gen.nextUnit();
        }
        for (int i = 0; i < N * N; i++) {
            b[i] = gen.nextUnit();
        }

        for (int r = 0; r < R; r++) {
            a[r] = a[r] + 1.0e-9; // anti-elimination, part of the algorithm
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    double s = 0.0;
                    for (int k = 0; k < N; k++) {
                        s = s + a[i * N + k] * b[k * N + j];
                    }
                    c[i * N + j] = s;
                }
            }
        }

        double sumC = 0.0;
        for (int i = 0; i < N * N; i++) {
            sumC = sumC + c[i];
        }

        return new Result(sumC, c[0], c[N * N - 1]);
    }

    public static void main(String[] args) {
        String mode = args.length > 0 ? args[0] : "once";
        int k = args.length > 1 ? Integer.parseInt(args[1]) : 7;

        Result res;
        if (mode.equals("steady")) {
            res = null;
            for (int i = 0; i < k; i++) {
                long t0 = System.nanoTime(); // section 5.2 monotonic clock for Java
                res = workload();
                long elapsed = System.nanoTime() - t0;
                System.out.println("ITER " + i + " " + elapsed);
                System.out.flush(); // section 5.2: flushed immediately
            }
        } else {
            res = workload();
        }

        System.out.printf(Locale.ROOT, "MB06 sumC=%.16e c_first=%.16e c_last=%.16e%n",
                res.sumC(), res.cFirst(), res.cLast());
    }
}

File: benchmark/2026-09-17-7677581(latest)/micro/src/kotlin/mb06.kt
// MB-06 - Matrix multiplication: classical i-j-k order over flat row-major arrays.

import java.util.Locale

/** Lehmer / MINSTD generator, frozen for every language in the suite. */
class Lcg(private var state: Long) {
    fun nextInt(): Long {
        state = (48271L * state) % 2147483647L
        return state
    }

    fun nextUnit(): Double = nextInt() / 2147483647.0
}

const val SEED = 20266917L
const val N = 512
const val R = 3

class Mb06Result(val sumC: Double, val cFirst: Double, val cLast: Double)

fun workload(): Mb06Result {
    // Section 5.2: the body re-seeds the generator and regenerates A and B,
    // so every steady iteration performs exactly the work `once` performs.
    val gen = Lcg(SEED)
    val a = DoubleArray(N * N)
    val b = DoubleArray(N * N)
    val c = DoubleArray(N * N)
    for (i in 0 until N * N) {
        a[i] = gen.nextUnit()
    }
    for (i in 0 until N * N) {
        b[i] = gen.nextUnit()
    }

    for (r in 0 until R) {
        a[r] = a[r] + 1.0e-9 // anti-elimination, part of the frozen algorithm
        for (i in 0 until N) {
            for (j in 0 until N) {
                var s = 0.0
                for (k in 0 until N) {
                    s = s + a[i * N + k] * b[k * N + j]
                }
                c[i * N + j] = s
            }
        }
    }

    var sumC = 0.0
    for (i in 0 until N * N) {
        sumC = sumC + c[i]
    }

    return Mb06Result(sumC, c[0], c[N * N - 1])
}

fun main(args: Array<String>) {
    // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
    val mode = if (args.isNotEmpty()) args[0] else "once"
    var res: Mb06Result
    if (mode == "steady") {
        val k = if (args.size > 1) args[1].toIntOrNull() ?: 7 else 7
        val iterations = if (k < 1) 7 else k
        res = Mb06Result(0.0, 0.0, 0.0)
        for (i in 0 until iterations) {
            val t0 = System.nanoTime() // section 5.2 pins System.nanoTime() for Kotlin
            res = workload()
            val elapsed = System.nanoTime() - t0
            println("ITER $i $elapsed")
            System.out.flush() // section 5.2: each ITER line is flushed immediately
        }
    } else {
        res = workload()
    }
    // Locale.ROOT: section 2.4 fixes '.' as the decimal separator.
    println(
        "MB06 sumC=%.16e c_first=%.16e c_last=%.16e"
            .format(Locale.ROOT, res.sumC, res.cFirst, res.cLast)
    )
}
